### PR TITLE
Issue 2 background

### LINKED ISSUE #2: Component: TextBackground
We need a component which can act as an `EmptySpace` but with a centered text (such as a logo).

A snippet to get started:
```java
package io.leonis.torch.component;

import com.googlecode.lanterna.*;
import com.googlecode.lanterna.gui2.*;
import java.awt.Color;
import lombok.*;

/**
 * The Class TextBackground
 *
 * This class is basically an {@link EmptySpace} with a text on top. The text is a tint brighter
 * than the background color and will be centered based on the current size of the {@link
 * com.googlecode.lanterna.screen.Screen}
 *
 * @author Jeroen de Jong
 */
@Value
@EqualsAndHashCode(callSuper = true)
public final class TextBackground extends AbstractComponent<TextBackground> {

  /**
   * Background label
   */
  private final String text;
  private final TextColor color;

  @Override
  protected ComponentRenderer<TextBackground> createDefaultRenderer() {
    return new ComponentRenderer<TextBackground>() {

      @Override
      public TerminalSize getPreferredSize(final TextBackground component) {
        return component.getParent().getSize();
      }

      @Override
      public void drawComponent(final TextGUIGraphics graphics, final TextBackground component) {
        // TODO this needs some cleanup; also the text should be centered
        graphics.applyThemeStyle(component.getThemeDefinition().getNormal());
        graphics.setBackgroundColor(component.getColor());
        graphics.fill(' ');

        final Label text = new Label(component.getText());
        text.setBackgroundColor(component.getColor());
        final Color brighterColor = component.getColor().toColor().brighter();
        text.setForegroundColor(
            new TextColor.RGB(brighterColor.getRed(), brighterColor.getGreen(),
                brighterColor.getBlue()));
        text.draw(graphics);
      }
    };
  }
}
```

### PR DESCRIPTION
**Pull request by ThomasHakkers**
fixes #2

### CODE REVIEW

**Comment by thumbnail** [comment]
Tip for the future; you can reference issues using the `#`-character followed by the issue number. such as https://help.github.com/articles/autolinked-references-and-urls/

**Comment by thumbnail** [comment]
Also adding a beautiful screenshot is always welcome:
image

Btw I replaced the bulk of your calculations by using `TextGraphics#putString`. It required manual looping which I think is a bit shitty though. Also, the position-calculation might be refactored into something more beautiful.

**Comment by thumbnail** [comment]
@romnous Since i also contributed; can you review the code?

File: src/main/java/io/leonis/torch/component/CenteredLabel.java [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+ */
+@Value
+public class CenteredLabel extends AbstractComponent<CenteredLabel> {
+  private final String text;
+  private final TextColor textColor;
+
+  @Override
+  protected synchronized TerminalSize calculatePreferredSize() {
+    return new TerminalSize(
+        Collections.max(this.getLines(), Comparator.comparing(String::length)).length(),
```
Comment (thumbnail) [inline src/main/java/io/leonis/torch/component/CenteredLabel.java]: Alternative;
```java
this.getLines().stream().mapToInt(String::length).max().orElse(0),
```

File: src/main/java/io/leonis/torch/component/TextBackground.java [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+ * @author Thomas Hakkers
+ */
+@Value
+@EqualsAndHashCode(callSuper = true)
+public final class TextBackground extends AbstractComponent<TextBackground> {
+
+  /**
+   * Background label
+   */
+  private final String text;
```
Comment (thumbnail) [inline src/main/java/io/leonis/torch/component/TextBackground.java]: Might be interesting to make `TextBackground` receive a Label / Component and rename it. It'd be more flexible and removes its dependency on `CenteredLabel`.

File: src/main/java/io/leonis/torch/component/CenteredLabel.java [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+        return component.getPreferredSize();
+      }
+
+      @Override
+      public void drawComponent(final TextGUIGraphics graphics, final CenteredLabel component) {
+        graphics.setForegroundColor(component.getTextColor());
+        IntStream.range(0, component.getLines().size())
+            .forEach(index -> {
+              graphics.putString(
+                  graphics.getSize().getColumns() / 2 - component.getPreferredSize().getColumns()/2,
```
Comment (romni) [inline src/main/java/io/leonis/torch/component/CenteredLabel.java]: factor out the `1/2`

File: src/main/java/io/leonis/torch/component/TextBackground.java [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+  private final TextColor backgroundColor;
+
+
```
Comment (romni) [inline src/main/java/io/leonis/torch/component/TextBackground.java]: redundant newline

File: src/main/java/io/leonis/torch/component/CenteredLabel.java [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+      public TerminalSize getPreferredSize(final CenteredLabel component) {
+        return component.getPreferredSize();
+      }
+
+      @Override
+      public void drawComponent(final TextGUIGraphics graphics, final CenteredLabel component) {
+        IntStream.range(0, component.getLines().size())
+            .forEach(index -> graphics.putString(
+                (graphics.getSize().getColumns() - component.getPreferredSize().getColumns())/2,
+                (graphics.getSize().getRows() - component.getPreferredSize().getRows())/2 + index,
```
Comment (RyanMeulenkamp) [inline src/main/java/io/leonis/torch/component/CenteredLabel.java]: Put spaces before and after the arithmetic operations

File: src/main/java/io/leonis/torch/component/ComponentBackground.java [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+public final class ComponentBackground extends AbstractComponent<ComponentBackground> {
+
+  private final Component component;
+  private final TextColor backgroundColor;
+
+  private TextColor getForegroundColor() {
+    return  new TextColor.RGB(
+        this.backgroundColor.toColor().brighter().getRed(),
+        this.backgroundColor.toColor().brighter().getGreen(),
+        this.backgroundColor.toColor().brighter().getBlue());
```
Comment (RyanMeulenkamp) [inline src/main/java/io/leonis/torch/component/ComponentBackground.java]: Put the closing parenthesis on the next line.

**Review by RyanMeulenkamp — CHANGES_REQUESTED** [review]

File: src/main/java/io/leonis/torch/component/ComponentBackground.java (hunk 6 above)
Comment (RyanMeulenkamp) [reply src/main/java/io/leonis/torch/component/ComponentBackground.java]: Also, there are two spaces after `return`

File: src/main/java/io/leonis/torch/component/CenteredLabel.java [hunk 7]
```diff
@@ -0,0 +1,51 @@
+package io.leonis.torch.component;
+
+import com.googlecode.lanterna.*;
+import com.googlecode.lanterna.gui2.*;
+import java.util.*;
+import java.util.stream.IntStream;
+import lombok.*;
+
+/**
+ * The Class CenteredLabel.
```
[the remark this answers is not in the document]
Comment (thumbnail) [reply src/main/java/io/leonis/torch/component/CenteredLabel.java:10]: This is a thing we do everywhere in the codebase. Might want to create an issue or discuss this on Slack if you want to change it.

**Review by RyanMeulenkamp — APPROVED** [review]

**Pull request merged**